### PR TITLE
Hclang 125

### PR DESCRIPTION
**Pull request by Abildgaard96**
Wrote a bit more for parser implementation (tried to atleast)

### Changelog
 - explained parsePotentialFunctionCall method
 - explained parseExpressionAtomic
     - basicly how the parser deals with expression..

### Issues solved 
add these with the respective key as HCLANG-1 (as an example)
 - HCLANG-125

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: report/4.Solution/ParserImplementation.tex [hunk 1]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
 The method then checks whether the new current symbol is an equals-symbol - if that's the case the declaration is followed by an expression - which means it's an declaration with an assignment otherwise the expression is set to null. 
 The method then returns a Declaration, which is a subclass of AstNode.  
 
+A more complex parsing routine is when an expression is to be parsed. 
+Since an expression in HCL may be either a functioncall or any kind of value, a lot of possibilities arise.
```
Comment (nillas12) [inline report/4.Solution/ParserImplementation.tex]: Pretty sure function call is two words.
I am also not a huge fan of "a lot of possibilites arise", I will think on it, don't have a good alternative at the moment.

File: report/4.Solution/ParserImplementation.tex [hunk 2]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+            if (symbol.isFunctions) {
+        	    ...parse needed arguments...
+        	    AstNode.Command.Expression.FunctionCall(
+        	    AstNode.Command.Expression.Value.Identifier(token.value),
+        	    listOf(expression) + secondaryArguments
+            } else expression
+        } else -> expression
+\end{lstlisting}
+
+As seen in snippet \ref{lis:parsePotCall}, the method will first determine if the current token is a lambda expression, signified by a \textit{blockStart-token} (3).
```
Comment (nillas12) [inline report/4.Solution/ParserImplementation.tex]: There are a lot of wills in the following paragraph, it gets a bit repetitive.
I know it is difficult to write differently, and it reads fine, so just a thought

File: report/4.Solution/ParserImplementation.tex [hunk 3]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
 The method then checks whether the new current symbol is an equals-symbol - if that's the case the declaration is followed by an expression - which means it's an declaration with an assignment otherwise the expression is set to null. 
 The method then returns a Declaration, which is a subclass of AstNode.  
 
+A more complex parsing routine is when an expression is to be parsed, since an expression in HCL may be either a function-call or any kind of value.
```
Comment (jones1994) [inline report/4.Solution/ParserImplementation.tex]: Change "a more complex parsing routine is when an expression is to be parsed" -> "It is slightly more complex to parse an expression".

(Maybe) Change "function-call or any kind of value" -> "function-call or any kind of evaluation to a value".

File: report/4.Solution/ParserImplementation.tex [hunk 4]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+        	    AstNode.Command.Expression.FunctionCall(
+        	    AstNode.Command.Expression.Value.Identifier(token.value),
+        	    listOf(expression) + secondaryArguments
+            } else expression
+        } else -> expression
+\end{lstlisting}
+
+As seen in snippet \ref{lis:parsePotCall}, the method first determines if the current token is a lambda expression, signified by a \textit{blockStart-token} (3).
+Should it be a lambda expression, the parser determines whether or not there is a functions call within the lambda that needs to be parsed as well (7).
+If there is not a function-call within the lambda, the lambda expression itself is returned from the method (10).
```
Comment (jones1994) [inline report/4.Solution/ParserImplementation.tex]: Change "Lambda" -> "Lambda expression" 
OR
Change "Lambda expression" -> "Lambda"

Comment (jones1994) [reply report/4.Solution/ParserImplementation.tex]: In the whole section offcause. Just to be consistent :)

File: report/4.Solution/ParserImplementation.tex [hunk 5]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+        	    AstNode.Command.Expression.Value.Identifier(token.value),
+        	    listOf(expression) + secondaryArguments
+            } else expression
+        } else -> expression
+\end{lstlisting}
+
+As seen in snippet \ref{lis:parsePotCall}, the method first determines if the current token is a lambda expression, signified by a \textit{blockStart-token} (3).
+Should it be a lambda expression, the parser determines whether or not there is a functions call within the lambda that needs to be parsed as well (7).
+If there is not a function-call within the lambda, the lambda expression itself is returned from the method (10). 
+Should the token be an identifier, the parser determines if the identifier signifies a function, in which case it is determined that a function-call is being made (16).
```
Comment (jones1994) [inline report/4.Solution/ParserImplementation.tex]: Change "signifies a function" -> "associated with a function"

File: report/4.Solution/ParserImplementation.tex [hunk 6]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+The method looks for the needed arguments for the function, and parse them as expressions.
+The method then returns the function-call with its corresponding arguments (18).
+If the identifier does not signify a function, it is then parsed as a regular expression.
```
Comment (jones1994) [inline report/4.Solution/ParserImplementation.tex]: If the my last comment leads to changes, this should be reflected in this sentence as well.

File: report/4.Solution/ParserImplementation.tex [hunk 7]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+        Token.Identifier -> {
+            val token = accept<Token.Identifier>()
+            retrieveSymbol(token.value).handle(..Handle identifier..)
+        }
+        else -> unexpectedTokenError(current.token)
+    }
+            
+\end{lstlisting}
 
+The method itself deals with determining the type of the expression that is to be parsed, and then parsing it atomically.
```
Comment (jones1994) [inline report/4.Solution/ParserImplementation.tex]: There is something going on with the tense on "parsing" in this sentence. Not sure what ;)

File: report/4.Solution/ParserImplementation.tex [hunk 8]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+All expressions that begin with the '\texttt{[}' character are determined to be list declarations and are parsed as such (3).\\
+Expressions beginning with a parentheses can either be a lambda declaration (6), a tuple (9) or simply an expression encapsulated in parenthesis (11).\\
+Literals are simply accepted (15).\\
+Expressions beginning with a colon is can only be functions passed as arguments, and are handled accordingly (16).\\
```
Comment (jones1994) [inline report/4.Solution/ParserImplementation.tex]: Remove "is"

File: report/4.Solution/ParserImplementation.tex [hunk 9]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+Identifiers can either signify values or function-calls.
```
Comment (jones1994) [inline report/4.Solution/ParserImplementation.tex]: Change "values" -> "variables"

File: report/4.Solution/ParserImplementation.tex [hunk 10]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+However, as potential function-calls have already been handled at this point, the identifier is determined to signify a value and can be parsed accordingly (21).
```
Comment (jones1994) [inline report/4.Solution/ParserImplementation.tex]: If my last comment leads to changes, it should be reflected in this sentence as well.

File: report/4.Solution/ParserImplementation.tex (hunk 6 above)
Comment (Lambcreek) [reply report/4.Solution/ParserImplementation.tex]: Jeg er ikke vild med at der står "regular expression" her.

File: report/4.Solution/ParserImplementation.tex (hunk 7 above)
Comment (Lambcreek) [reply report/4.Solution/ParserImplementation.tex]: Enig med Jonas. Ændr til
"The method determines ..... and then parses it ....."

File: report/4.Solution/ParserImplementation.tex [hunk 11]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+            retrieveSymbol(token.value).handle(..Handle identifier..)
+        }
+        else -> unexpectedTokenError(current.token)
+    }
+            
+\end{lstlisting}
 
+The method itself deals with determining the type of the expression that is to be parsed, and then parsing it atomically.
+All expressions that begin with the '\texttt{[}' character are determined to be list declarations and are parsed as such (3).\\
+Expressions beginning with a parentheses can either be a lambda declaration (6), a tuple (9) or simply an expression encapsulated in parenthesis (11).\\
```
Comment (Lambcreek) [inline report/4.Solution/ParserImplementation.tex]: "parenthesis" skal være "parentheses"

File: report/4.Solution/ParserImplementation.tex [hunk 12]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+            } else {
+                val lambda = parseLambdaBody(emptyList())
+                AstNode.Command.Expression.LambdaExpression(listOf(), lambda.type, lambda.lambdaBody)
+            }
+        is Token.Identifier -> {
+            val token = accept<Token.Identifier>()
+            val symbol = retrieveSymbol(token.value)
+            if (symbol.isFunctions) {
+        	    ...parse needed arguments...
+        	    AstNode.Command.Expression.FunctionCall(
```
Comment (C0DK) [inline report/4.Solution/ParserImplementation.tex]: open bracket uden closing bracket

File: report/4.Solution/ParserImplementation.tex (hunk 6 above)
Comment (C0DK) [reply report/4.Solution/ParserImplementation.tex]: hvorfra får du regular expresssion anyway? det ser jeg ikke i koden.

File: report/4.Solution/ParserImplementation.tex [hunk 13]
```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
...
+                } else {
+                    parseExpression()
+                }
+            }
+        }
+        Token.Literal -> acceptLiteral()
+        Token.SpecialChar.Colon -> {
+        	accept<Token.SpecialChar.Colon>()
+        	val token = accept<Token.Identifier>()
+        	retrieveSymbol(token.value).handle(..handle token as function..)
```
Comment (C0DK) [inline report/4.Solution/ParserImplementation.tex]: din punktum løsning skal måske reflekteres over det hele... og samtidig måske gøre sådan at de er på egen linje så det ikke ligner noget kode eller ligende.. evt gør det med comments

File: report/4.Solution/ParserImplementation.tex (hunk 11 above)
Comment (C0DK) [inline report/4.Solution/ParserImplementation.tex]: behøver vi en så "atomisk" forklaring? det er ikke for meget?

File: report/4.Solution/ParserImplementation.tex (hunk 9 above)
Comment (C0DK) [reply report/4.Solution/ParserImplementation.tex]: remove "-"

Comment (C0DK) [reply report/4.Solution/ParserImplementation.tex]: or not? idk...

File: report/4.Solution/ParserImplementation.tex (hunk 6 above)
Comment (nillas12) [reply report/4.Solution/ParserImplementation.tex]: Tror der menes at det er en "regular" expression, ikke en regex.

**Pull request merged**